### DISCUSSION TITLE
JaversRepository implementation for Redis

### DISCUSSION THREAD

**Discussion by assen-sharlandjiev-akros-ch** [javers/javers#1394] [Ideas]
Hi,

I'd like to propose a new feature - implementation of JaversRepository for Redis and Spring Boot auto configuration for that. 
I'm also ready to start working on that feature and contribute it when ready. 

@bartoszwalacik Please let me know if that's okay with you?

**Comment by bartoszwalacik** [comment]
Sure, go on, let me know if you need some hints from me

**Comment by assen-sharlandjiev-akros-ch** [comment]
Hi, 
I have made some progress. 
If you want to take a look, the fork is [here](https://github.com/assen-sharlandjiev-akros-ch/javers/tree/feature/1394_redis_repository).
Should I create an issue (new feature tag) to track the commits? 

Cheers,
Assen.

**Reply by bartoszwalacik** [reply]
Hi @assen-sharlandjiev-akros-ch , a new issue is not needed, we can treat this discussion as an issue. Could you please create a PR from your fork to Javers master?

**Reply by bartoszwalacik** [reply]
btw could you please explain the design behind CdoSnapshotKeyExpireListener?

**Reply by assen-sharlandjiev-akros-ch** [reply]
Hi. I'll open the PR, but please consider it's not ready for merge yet.

Let me give you some context about your question:

With Redis, one can specify key expiration (duration). Once a key expires Redis will remove it from the database. 
We store the Javers Snapshots for a given object under Redis lists, whose key is `jv_snapshots:<entity class name>/<object id>`.
The JaversRedisRepository constructor takes a Duration parameter to specify the key expiration period. In this way if we provide a positive value for the duration, Redis will perform housekeeping and will remove expired snotshot data. 

Some of the queries defined in the RedisRepository interface require the retrieval of CDO snapshot data from multiple keys. 

For example  the result for `List<CdoSnapshot> getStateHistory(final Set<ManagedType> givenClasses, final QueryParams queryParams)`  will require to access the CDO snapshots list for each key matching a pattern: `jv_snapshots:<entity class name>/*`. 

In general key scan operations are expensive, and in large Redis databases it takes some time to obtain the result. To optimise the response time, and to lower the load on the Redis service, we are using Reids sets, to store the CDO snapshot list keys relevant to a given entity. These sets have keys with the following pattern:  `jv_snapshots_keys:<entity class name>`. Besides that we also have a set with key `jv_snapshots_keys` that holds all of the Javers Snapshots list keys . We use these sets in the query methods, in place of key scans. 

Now the answer to your question - what does the `CdoSnapshotKeyExpireListener` do and why do we need it:

Consider that a given key, for example, `jv_snapshots:org.javers.repository.redis.domain.Device/thermometer` expires, and it is removed by Redis. At the same time, we already have that key stored in two sets - `jv_snapshots_keys` and `jv_snapshots_keys:org.javers.repository.redis.domain.Device`. These set entries are not removed from the sets automatically by Redis. 

Redis, on the other hand, provides key space notifications, e.g. we can process an event on a specified key pattern, when it occurs. The CdoSnapshotKeyExpireListener is precisely that, an event listener on key expiration events, whose responsibility is to remove the expired key entries from the relevant key sets.  When an instance of the JaversRedisRepository is created, a subscription for key space events with `__key*__:jv_snapshots:*` pattern is created, and events are processed by the CdoSnapshotKeyExpireListener.

Now there's a "catch" with this implementation which we are currently working on. Consider that Redis keys may expire and be removed by the Redis server while there's no active CdoSnapshotKeyExpireListener running. In this case, the expired key entries will remain in the sets. We should be able to address this problem, by implementing similar key listener, that resides and executes on the Redis server itself.  Will post here, as soon as we have it.

**Reply by bartoszwalacik** [reply]
got it, thanks for your explanation @assen-sharlandjiev-akros-ch , please keep me posted

**Reply by bartoszwalacik** [reply]
btw I wonder what's is your use case, I mean why did you decide that you need RedisRepository for Javers. Do you need to track only a last few days of an object history, something like a short-term memory of what happened recently?

**Reply by assen-sharlandjiev-akros-ch** [reply]
Hi, there are multiple reasons. 
I have project that use only Redis for persistence. For example, we store some "configuration" data in Redis database. For the sake of simplicity It makes sense, to use key-value database, rather than RDBMS and ORM. A common requirement is to be able to keep track of changes on these "configurations", and who made those changes, and when. So in this scenario Javers + Redis + Spring Data for Redis makes perfect sense. Also having automated housekeeping and retention for the audit metadata makes it easier for operations. In some scenarios  I don't need to store audit metadata for long periods.

On the other hand, there's a project with quite big amount of data in RDBMS and that data changes frequently, which leads to huge tables for Javers. I'd like to be able to store and update domain objects in RDBMS, but at the same time, I'd like to be able to configure Spring Data for JPA to use Javers with JaversRedisRepository and store the audit data in Redis. this will unload the SQL server. 
Hope that this makes sense.

**Reply by bartoszwalacik** [reply]
thanks for the explanation

**Reply by IvanSimeonov** [reply]
Hi @bartoszwalacik, 

Quick update on our progress:

We've submitted a new pull request for the feature - implementation of JaversRepository for Redis and Spring Boot auto-configuration. You can check it out [here.](https://github.com/javers/javers/pull/1403)

Additionally, we've created a new fork of the documentation repository to include updates for this feature. If you have any suggestions, please let us know.

Cheers,
Ivan

**Reply by bartoszwalacik** [reply]
Hi @IvanSimeonov , ok, thanks, I'm on holiday now, will take a look on it in 1-2 weeks

**Reply by IvanSimeonov** [reply]
Hi @bartoszwalacik,

I wanted to provide a quick update on our progress:

We have made some changes in the JaVers Redis AutoConfiguration. Additionally, we've submitted a new PR to update the JaVers Redis support documentation. You can review it [here.](https://github.com/javers/javers.github.io/pull/125)

If you have any suggestions or necessary improvements, we'd love to hear them.

Best Regards,
Ivan

**Reply by IvanSimeonov** [reply]
Hi @bartoszwalacik,

We've created a fork of the [Spring PetClinic](https://github.com/spring-projects/spring-petclinic) project to showcase how JaVers can be used with Redis. You can find the project [here](https://github.com/IvanSimeonov/spring-petclinic/tree/feature/javers-redis-demo). We'd appreciate it if you could take a look and confirm that the new Redis support is working as expected.

If you have any feedback or suggestions for improvements, we'd love to hear from you.

Thanks in advance,
Ivan

**Reply by bartoszwalacik** [reply]
ok @IvanSimeonov , I will take a look in this week

**Reply by bartoszwalacik** [reply]
Hi @IvanSimeonov,

Please find my comments in the PR. Your work looks really good! I did a quick review to ensure that the code aligns with the conventions in the Javers repo. Since I'm not familiar with Redis, I don't want to dive too deeply into your PR. Will you be ready to merge after applying my comments?

Your contribution is significant! Do you think you could take ownership of this new module in terms of future maintenance? You could simply take on the committer role for the javers-redis module.

**Comment by IvanSimeonov** [comment]
Hi @bartoszwalacik,

Thank you for your review and kind words! We will address the comments and aim to merge the PR after that. However, since we don't have much experience with Spock, it might take us some time to fully address the comments properly.

Both of us are happy to take on the responsibility of becoming contributors/maintainers, and we're excited to support the javers-redis module moving forward.

Regarding the documentation, could you take a quick look at the PR to see if any improvements are needed and provide feedback?

Thanks again for your support!

**Reply by bartoszwalacik** [reply]
Thank you very much for your willingness to become a maintainer of the Redis module 👍 

Let me know if you need any support with writing the new Spock test. Spock is super easy to use, so it shouldn't be too much work for you. Simply extend the abstract E2E test I pointed out, add the Redis repository configuration, and all the tests from the abstract class will be executed against the Javers instance with the Redis repository.

Regarding the documentation, LGTM, I left some minor comments. thanks

**Reply by assen-sharlandjiev-akros-ch** [reply]
Thank you for asking us to implement the Spock tests. 

This helped us to realise that we have some bugs in the Reids repository implementation that we need to address. 

For example how we handle the unbounded value objects, or how we applied the limits from the query params.

Will ping you as soon as we resolve these for a quick review.

**Reply by bartoszwalacik** [reply]
awesome, see also my comment with a link to the build https://github.com/javers/javers/pull/1403#issuecomment-2383534595

**Comment by IvanSimeonov** [comment]
Hello @bartoszwalacik ,

I am following up in case you missed my message in the pull request.

All work related to the Redis functionality has been completed and included in both pull requests—the documentation and main repositories.

Please let us know if you have any feedback or suggestions. We’re happy to make any adjustments if needed.

Cheers,
Ivan

**Reply by bartoszwalacik** [reply]
Hi @IvanSimeonov , I didn't miss :) I'm merging and releasing it right now.
v 7.8.0
The new artifacts should be available in the Central in a few hours

Thanks for the great contribution!

see 
 https://javers.org/release-notes
 https://javers.org/
